src/files.c

```diff
@@ -0,0 +1,361 @@
+		return false;
+	}
+
+	strcpy(path_buffer, original_path);
+	strcat(path_buffer, extension);
+
+	return true;
+}
+
+bool 
+file_is_compressed_type(char const *path)
+{
+	int len = (int)strlen(path);
+
+	if (strcmp(path + len - 3, ".gz") == 0 || strcmp(path + len - 3, "-gz") == 0) {
+		return true;
+	} else if (strcmp(path + len - 2, ".z") == 0 || strcmp(path + len - 2, "-z") == 0 || strcmp(path + len - 2, "_z") == 0 || strcmp(path + len - 2, ".Z") == 0) {
+		return true;
+	}
+
+	return false;
+}
+
+void
+files_shutdown()
+{
+	struct x16file *f = open_files;
+	struct x16file *next_f = NULL;
+	for(; f != NULL; f = next_f) {
+		next_f = f->next;
+		x16close(f);
+	}
+}
+
+struct x16file *
+x16open(const char *path, const char *attribs)
+{
+	struct x16file *f = malloc(sizeof(struct x16file));
+	strcpy(f->path, path);
+
+	if(file_is_compressed_type(path)) {
+		char tmp_path[PATH_MAX];
+		if(!get_tmp_name(tmp_path, path, ".tmp")) {
+			printf("Path too long, cannot create temp file: %s\n", path);
+			goto error;
+		}
+
+		gzFile zfile = gzopen(path, "rb");
+		if(zfile == Z_NULL) {
+			printf("Could not open file for decompression: %s\n", path);
+			goto error;
+		}
+
+		SDL_RWops *tfile = SDL_RWFromFile(tmp_path, "wb");
+		if(tfile == NULL) {
+			gzclose(zfile);
+			printf("Could not open file for write: %s\n", tmp_path);
+			goto error;
+		}
+
+		printf("Decompressing %s\n", path);
+
+		const int buffer_size = 16 * 1024 * 1024;
+		char *buffer = malloc(buffer_size);
+
+		int read = gzread(zfile, buffer, buffer_size);
+		int64_t total_read = read;
+		const int64_t progress_increment = 128 * 1024 * 1024;
+		int64_t progress_threshold = progress_increment;
+		while(read > 0) {
+			if(total_read > progress_threshold) {
+				printf("%" PRId64 " MB\n", total_read / (1024 * 1024));
+				progress_threshold += progress_increment;
+			}
+			SDL_RWwrite(tfile, buffer, read, 1);
+			read = gzread(zfile, buffer, buffer_size);
+			total_read += read;
+		}
+		printf("%" PRId64 " MB\n", total_read / (1024 * 1024));
+
+		SDL_RWclose(tfile);
+		gzclose(zfile);
+		free(buffer);
+
+		f->file = SDL_RWFromFile(tmp_path, attribs);
+		if(f->file == NULL) {
+			unlink(tmp_path);
+			goto error;
+		}
+		f->size = total_read;
+	} else {
+		f->file = SDL_RWFromFile(path, attribs);
+		if(f->file == NULL) {
+			goto error;
+		}
+		f->size = SDL_RWsize(f->file);
+	}
+	f->pos = 0;
+	f->modified = false;
+	f->next = open_files ? open_files : NULL;
+	open_files = f;
+
+	return f;
+
+error:
+	free(f);
+	return NULL;
+}
+
+void 
+x16close(struct x16file *f)
+{
+	if(f == NULL) {
+		return;
+	}
+
+	SDL_RWclose(f->file);
+
+	if(file_is_compressed_type(f->path)) {
+		char tmp_path[PATH_MAX];
+		if(!get_tmp_name(tmp_path, f->path, ".tmp")) {
+			printf("Path too long, cannot create temp file: %s\n", f->path);
+			goto tmp_path_error;
+		}
+
+		if(f->modified == false) {
+			goto zfile_clean;
+		}
+
+		gzFile zfile = gzopen(f->path, "wb9");
```

Comment:
gzip 9 is suuuuuper slow.  Perhaps consider 6 or 1?

Reply:
I'm fine with a lower value. I guess I didn't want to re-compress a gzipped file into something even smaller, and it's not clear how to determine the compression rate of a file since zlib doesn't appear to expose that information through their file I/O interface.

Reply:
Just a consideration, I'm not objecting.  For reasonably-sized zipped SD cards, it's unlikely to matter much.

Reply:
It's a fine consideration. I'll go ahead and go with 6 as hopefully a good compromise between speed and compression.